File: engine-wrapper/engine-wrapper.py
import asyncio
import os
import sys
import logging
import subprocess
import json
from logging.handlers import RotatingFileHandler
from datetime import datetime, timezone
from dotenv import load_dotenv
from pathlib import Path

# Determine if running as a compiled executable (Nuitka or PyInstaller)
is_compiled_or_frozen = getattr(sys, 'frozen', False) or '__compiled__' in globals()

if is_compiled_or_frozen:
    # Running as compiled executable
    # Nuitka --onefile might set sys.executable to temp dir, so use sys.argv[0]
    script_dir = Path(sys.argv[0]).resolve().parent
else:
    # Running as a script
    script_dir = Path(__file__).resolve().parent

# Configure logging
log_handlers = []
# If running as a compiled executable, always log to file to ensure stability in non-interactive environments
# (e.g. Task Scheduler, Services) where sys.stderr might be invalid or cause crashes (0x80070001).
if is_compiled_or_frozen:
    log_file = script_dir / 'engine-wrapper.log'
    log_handlers.append(RotatingFileHandler(
        log_file, 
        maxBytes=1*1024*1024, 
        backupCount=2, 
        encoding='utf-8'
    ))
else:
    # Development mode: Log to console
    log_handlers.append(logging.StreamHandler())

logging.basicConfig(
    level=logging.INFO,
    format='[%(asctime)s.%(msecs)03dZ] %(message)s',
    datefmt='%Y-%m-%dT%H:%M:%S',
    handlers=log_handlers
)
logging.Formatter.converter = lambda *args: datetime.now(timezone.utc).timetuple()

logging.debug(f"is_compiled_or_frozen: {is_compiled_or_frozen}")
logging.debug(f"sys.executable: {sys.executable}")
logging.debug(f"sys.argv[0]: {sys.argv[0]}")
logging.debug(f"Resolved script_dir: {script_dir}")

load_dotenv(dotenv_path=script_dir / '.env')

HOST = os.getenv('BIND_ADDRESS', '127.0.0.1')
PORT = int(os.getenv('LISTEN_PORT', '4082'))

def get_engine_list():
    engines_json_path = script_dir / 'engines.json'
    engines = []
    
    if engines_json_path.exists():
        try:
            with open(engines_json_path, 'r', encoding='utf-8') as f:
                engines = json.load(f)
        except Exception as e:
            logging.error(f"Failed to parse engines.json: {e}")
    else:
        logging.error(f"engines.json not found at {engines_json_path}. No engines available.")
        
    return engines

async def pipe_stream(reader: asyncio.StreamReader, writer: asyncio.StreamWriter, log_prefix: str):
    try:
        while not reader.at_eof():
            data = await reader.read(1024)
            if not data:
                break
            
            text = data.decode(errors='ignore').strip()
            # Reduce logging noise: Skip 'info' commands unless debugging
            if text.startswith("info"):
                logging.debug(f"{log_prefix} {text}")
            else:
                logging.info(f"{log_prefix} {text}")

            writer.write(data)
            await writer.drain()
    except (ConnectionResetError, BrokenPipeError, asyncio.CancelledError, ConnectionAbortedError):
        pass
    except Exception as e:
        logging.error(f"Unexpected error in {log_prefix}: {e}", exc_info=True)
    finally:
        pass

async def apply_engine_options(stdin: asyncio.StreamWriter, options: dict):
    """Apply engine options from engines.json configuration."""
    if not options or not isinstance(options, dict):
        return
    
    for name, value in options.items():
        # Normalize boolean values to lowercase 'true'/'false' for USI compatibility
        if isinstance(value, bool):
            value = str(value).lower()

        command = f"setoption name {name} value {value}\n"
        logging.info(f"Applying option: {command.strip()}")
        
        try:
            stdin.write(command.encode())
            await stdin.drain()
        except (BrokenPipeError, ConnectionResetError) as e:
            logging.error(f"Failed to write option '{name}': {e}")
            raise
        except Exception as e:
            logging.error(f"Unexpected error writing option '{name}': {e}", exc_info=True)
            raise

async def handle_client(client_reader: asyncio.StreamReader, client_writer: asyncio.StreamWriter):
    peername = client_writer.get_extra_info('peername')
    logging.info(f"Client connected from {peername}")
    engine_process = None
    tasks_to_cancel = []

    try:
        first_line = await client_reader.readline()
        if not first_line:
            logging.warning("Client disconnected before sending command.")
            return

        command_line = first_line.decode().strip()
        logging.info(f"Received command: '{command_line}'")

        engines = get_engine_list()

        if command_line == 'list':
            list_response = json.dumps(engines)
            client_writer.write(list_response.encode() + b'\n')
            await client_writer.drain()
            client_writer.close()
            await client_writer.wait_closed()
            return

        engine_id = ''
        if command_line.startswith('run '):
            engine_id = command_line[4:].strip()
        elif command_line == 'research' or command_line == 'game':
            # Backward compatibility
            engine_id = command_line
        else:
            logging.error(f"Invalid command received: {command_line}")
            client_writer.write(b"WRAPPER_ERROR: Invalid command. Use 'list' or 'run <id>'.\n")
            await client_writer.drain()
            return

        engine_def = next((e for e in engines if e['id'] == engine_id), None)
        if not engine_def:
            logging.error(f"Engine ID '{engine_id}' not found.")
            client_writer.write(f"WRAPPER_ERROR: Engine ID '{engine_id}' not found.\n".encode())
            await client_writer.drain()
            return

        engine_path_str = engine_def.get('path')
        if not engine_path_str:
            logging.error(f"Engine path for ID '{engine_id}' is not set.")
            client_writer.write(b"WRAPPER_ERROR: Engine path configuration error.\n")
            await client_writer.drain()
            return
        
        engine_path = Path(engine_path_str)
        # Resolve relative paths relative to the script directory
        if not engine_path.is_absolute():
            engine_path = (script_dir / engine_path).resolve()

        engine_directory = engine_path.parent

        # Prevent new console window on Windows
        creationflags = 0
        if sys.platform == "win32":
            creationflags = subprocess.CREATE_NO_WINDOW

        try:
            engine_process = await asyncio.create_subprocess_exec(
                str(engine_path),
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=engine_directory,
                creationflags=creationflags
            )
        except FileNotFoundError:
            logging.error(f"Engine executable not found at '{engine_path}'")
            error_message = "WRAPPER_ERROR: Engine executable not found."
            client_writer.write(error_message.encode() + b'\n')
            await client_writer.drain()
            return
        except Exception as e:
            logging.error(f"Failed to start engine process: {e}", exc_info=True)
            error_message = "WRAPPER_ERROR: Failed to start engine process."
            client_writer.write(error_message.encode() + b'\n')
            await client_writer.drain()
            return

        logging.info(f"Started engine process: {engine_path} (PID: {engine_process.pid})")

        options_applied = False  # Track if options have been applied

        async def client_to_engine():
            nonlocal options_applied
            try:
                while True:
                    line_bytes = await client_reader.readline()
                    if not line_bytes:
                        break
                    command = line_bytes.decode().strip()
                    
                    # Inject options immediately BEFORE 'isready' command (only once)
                    if command == 'isready' and not options_applied:
                        options = engine_def.get('options')
                        if options:
                            logging.info(f"Detected 'isready', applying engine options for '{engine_id}'...")
                            await apply_engine_options(engine_process.stdin, options)
                            options_applied = True

                    logging.info(f"[Client -> Engine] {command}")
                    engine_process.stdin.write(line_bytes)
                    await engine_process.stdin.drain()
            except Exception as e:
                logging.debug(f"Client to engine pipe closed: {e}")

        client_to_engine_task = asyncio.create_task(client_to_engine())
        engine_stdout_to_client_task = asyncio.create_task(
            pipe_stream(engine_process.stdout, client_writer, "[Engine -> Client]")
        )
        engine_stderr_to_client_task = asyncio.create_task(
            pipe_stream(engine_process.stderr, client_writer, "[Engine ERROR]")
        )
        engine_wait_task = asyncio.create_task(engine_process.wait())
        
        tasks_to_cancel = [
            client_to_engine_task, 
            engine_stdout_to_client_task, 
            engine_stderr_to_client_task, 
            engine_wait_task
        ]

        done, pending = await asyncio.wait(
            tasks_to_cancel,
            return_when=asyncio.FIRST_COMPLETED
        )

        for task in pending:
            task.cancel()

    except Exception as e:
        logging.error(f"An error occurred in client handler: {e}", exc_info=True)
    finally:
        for task in tasks_to_cancel:
            if not task.done():
                task.cancel()

        if engine_process:
            if engine_process.returncode is not None:
                logging.info(f"Engine process (PID: {engine_process.pid}) already exited with code {engine_process.returncode}.")
            else:
                logging.info(f"Cleaning up engine process (PID: {engine_process.pid}).")
                try:
                    # Send 'quit' command
                    if engine_process.stdin and not engine_process.stdin.is_closing():
                        logging.info("Sending 'quit' command to engine.")
                        engine_process.stdin.write(b'quit\n')
                        await engine_process.stdin.drain()
                        engine_process.stdin.close()
                    
                    # Wait for engine to exit
                    try:
                        await asyncio.wait_for(engine_process.wait(), timeout=5.0)
                        logging.info(f"Engine process (PID: {engine_process.pid}) exited gracefully.")
                    except asyncio.TimeoutError:
                        logging.warning(f"Engine did not exit after 'quit' command. Terminating.")
                        if engine_process.returncode is None:
                            engine_process.terminate()
                            try:
                                await asyncio.wait_for(engine_process.wait(), timeout=3.0)
                            except asyncio.TimeoutError:
                                logging.warning(f"Engine process (PID: {engine_process.pid}) did not terminate gracefully, killing.")
                                if engine_process.returncode is None:
                                    engine_process.kill()
                                    await engine_process.wait()
                except (BrokenPipeError, ConnectionResetError, ConnectionAbortedError):
                    logging.warning("Engine stdin pipe already closed, could not send 'quit'.")
                except ProcessLookupError:
                    pass
            engine_process = None

        if client_writer and not client_writer.is_closing():
            client_writer.close()
            try:
                await client_writer.wait_closed()
            except Exception:
                pass
        
        logging.info(f"Client disconnected from {peername}.")


async def main():
    server = await asyncio.start_server(handle_client, HOST, PORT)
    addrs = ', '.join(str(sock.getsockname()) for sock in server.sockets)
    logging.info(f"Single-port engine wrapper server listening on {addrs}")
    
    engines_json_path = script_dir / 'engines.json'
    if engines_json_path.exists():
        logging.info(f"engines.json found at {engines_json_path}")
        try:
            with open(engines_json_path, 'r', encoding='utf-8') as f:
                engines = json.load(f)
            logging.info(f"Loaded {len(engines)} engines from engines.json:")
            for e in engines:
                logging.info(f"  - {e.get('id')}: {e.get('name')} ({e.get('path')})")
        except Exception as e:
            logging.error(f"Failed to parse engines.json: {e}")
    else:
        logging.error(f"engines.json not found. Please create one based on engines.json.example.")

    async with server:
        await server.serve_forever()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logging.info("Server is shutting down.")

File: engine-wrapper/engine-wrapper.mjs
import { spawn } from 'child_process';
import net from 'net';
import readline from 'readline';
import dotenv from 'dotenv';
import path from 'path';
import fs from 'fs';
import crypto from 'crypto';
import { fileURLToPath } from 'url';

// Find .env file in the same directory as this script
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.join(__dirname, '.env') });

const HOST = process.env.BIND_ADDRESS || '127.0.0.1';
const PORT = parseInt(process.env.LISTEN_PORT || '4082', 10);
const ACCESS_TOKEN = process.env.WRAPPER_ACCESS_TOKEN;

/**
 * Load engine list from engines.json.
 */
function getEngineList() {
  const enginesJsonPath = path.join(__dirname, 'engines.json');
  let engines = [];

  if (fs.existsSync(enginesJsonPath)) {
    try {
      const content = fs.readFileSync(enginesJsonPath, 'utf-8');
      engines = JSON.parse(content);
    } catch (e) {
      console.error(`[${new Date().toISOString()}] Failed to parse engines.json: ${e.message}`);
    }
  } else {
    console.error(`[${new Date().toISOString()}] engines.json not found at ${enginesJsonPath}. No engines available.`);
  }

  return engines;
}

/**
 * Apply engine options from engines.json configuration.
 * Sends setoption commands before 'isready'.
 */
function applyEngineOptions(engineProcess, options) {
  if (!options || typeof options !== 'object') {
    return;
  }

  for (const [name, value] of Object.entries(options)) {
    const command = `setoption name ${name} value ${value}`;
    console.log(`[${new Date().toISOString()}] Applying option: ${command}`);
    
    if (engineProcess && engineProcess.stdin && engineProcess.stdin.writable) {
      try {
        engineProcess.stdin.write(command + '\n', (err) => {
          if (err) {
            console.error(`[${new Date().toISOString()}] Failed to write option '${name}': ${err.message}`);
          }
        });
      } catch (e) {
        console.error(`[${new Date().toISOString()}] Exception writing option '${name}': ${e.message}`);
      }
    } else {
      console.warn(`[${new Date().toISOString()}] Cannot apply option '${name}': stdin not writable`);
    }
  }
}

const server = net.createServer((socket) => {
  console.log(`[${new Date().toISOString()}] Client connected.`);
  let engineProcess = null;
  let isCleaningUp = false;
  let optionsApplied = false; // Track if options have been applied
  let authenticated = !ACCESS_TOKEN; // If no token set, auth is not required
  let engineStarted = false; // Track if engine process is running
  let authNonce = null;

  const cleanup = () => {
    if (isCleaningUp) {
      return;
    }
    isCleaningUp = true;

    // Stop reading from the socket
    if (rl) {
      rl.close();
    }

    // If engine process is already gone or exited
    if (!engineProcess || engineProcess.exitCode !== null) {
      engineProcess = null;
      engineStarted = false;
      if (!socket.destroyed) {
        console.log(`[${new Date().toISOString()}] Closing client socket.`);
        socket.destroy();
      }
      return;
    }

    console.log(`[${new Date().toISOString()}] Cleaning up engine process (PID: ${engineProcess.pid}).`);

    let termTimeout;
    const quitTimeout = setTimeout(() => {
      console.warn(`[${new Date().toISOString()}] Engine did not exit after 'quit'. Terminating.`);
      if (engineProcess) engineProcess.kill('SIGTERM');

      termTimeout = setTimeout(() => {
        console.warn(`[${new Date().toISOString()}] Engine did not respond to SIGTERM. Killing.`);
        if (engineProcess) engineProcess.kill('SIGKILL');
      }, 3000);
    }, 5000);

    engineProcess.once('close', (code) => {
      clearTimeout(quitTimeout);
      clearTimeout(termTimeout);
      console.log(`[${new Date().toISOString()}] Engine process exited with code ${code}.`);
      engineProcess = null;
      // Do not reset isCleaningUp to false, as this connection scope is done.
      if (!socket.destroyed) {
        socket.destroy();
      }
    });

    try {
      if (engineProcess.stdin && engineProcess.stdin.writable) {
        console.log(`[${new Date().toISOString()}] Sending 'quit' command to engine.`);
        engineProcess.stdin.write('quit\n');
        engineProcess.stdin.end();
      }
    } catch (e) {
      console.error(`[${new Date().toISOString()}] Failed to send 'quit' command, proceeding to terminate.`, e.message);
    }
  };

  if (!authenticated) {
    authNonce = crypto.randomBytes(16).toString('hex');
    socket.write(`auth_cram_sha256 ${authNonce}\n`);
  }

  const rl = readline.createInterface({ input: socket });

  rl.on('line', (line) => {
    // If engine is started, forward everything to it
    if (engineStarted) {
      if (engineProcess && engineProcess.stdin.writable) {
        const command = line; 
        const cmd = command.trim();

        // Inject options immediately BEFORE 'isready' command (only once)
        if (cmd === 'isready' && !optionsApplied && socket.engineOptions) {
          console.log(`[${new Date().toISOString()}] Detected 'isready', applying engine options...`);
          applyEngineOptions(engineProcess, socket.engineOptions);
          optionsApplied = true;
        }
        
        console.log(`[Client -> Engine] ${cmd}`);
        engineProcess.stdin.write(cmd + '\n');
      }
      return;
    }

    const input = line.trim();

    if (!authenticated) {
      if (input.startsWith('auth ')) {
        const digest = input.substring(5).trim();
        const expectedDigest = crypto.createHmac('sha256', ACCESS_TOKEN).update(authNonce).digest('hex');
        const digestBuffer = Buffer.from(digest, 'hex');
        const expectedDigestBuffer = Buffer.from(expectedDigest, 'hex');

        // Check length first to avoid RangeError in timingSafeEqual (DoS protection)
        // Then use timing-safe comparison to prevent timing attacks
        if (digestBuffer.length === expectedDigestBuffer.length && 
            crypto.timingSafeEqual(digestBuffer, expectedDigestBuffer)) {
          console.log(`[${new Date().toISOString()}] Client authenticated successfully.`);
          authenticated = true;
          socket.write('auth_ok\n');
          return;
        } else {
          console.warn(`[${new Date().toISOString()}] Authentication failed.`);
          socket.write('WRAPPER_ERROR: Authentication failed\n', () => socket.destroy());
        }
      } else {
        console.warn(`[${new Date().toISOString()}] Unauthenticated command attempt: ${input}`);
        socket.write('WRAPPER_ERROR: Authentication required\n', () => socket.destroy());
      }
      
      // Stop processing any further input immediately
      if (rl) {
        rl.close();
        rl.removeAllListeners();
      }
      return;
    }
    
    // Engine started check again just in case (though we checked at top)
    if (engineStarted) {
       // Logic moved to top of listener
       return;
    }

    console.log(`[${new Date().toISOString()}] Received command: '${input}'`);

    const engines = getEngineList();

    if (input === 'list') {
      const listResponse = JSON.stringify(engines);
      socket.write(listResponse + '\n');
      socket.end();
      return;
    }

    let engineId = '';
    if (input.startsWith('run ')) {
      engineId = input.substring(4).trim();
    } else if (input === 'research' || input === 'game') {
      // Backward compatibility
      engineId = input;
    } else {
      console.error(`[${new Date().toISOString()}] Invalid command received: ${input}`);
      socket.write(`WRAPPER_ERROR: Invalid command. Use 'list' or 'run <id>'.\n`);
      cleanup();
      return;
    }

    const engineDef = engines.find(e => e.id === engineId);
    if (!engineDef) {
      console.error(`[${new Date().toISOString()}] Engine ID '${engineId}' not found.`);
      socket.write(`WRAPPER_ERROR: Engine ID '${engineId}' not found.\n`);
      cleanup();
      return;
    }

    let enginePath = engineDef.path;
    if (!enginePath) {
      console.error(`[${new Date().toISOString()}] Engine path for ID '${engineId}' is not set.`);
      socket.write(`WRAPPER_ERROR: Engine path configuration error.\n`);
      cleanup();
      return;
    }

    // Resolve relative paths relative to the script directory to ensure 'cwd' is absolute and correct
    if (!path.isAbsolute(enginePath)) {
      enginePath = path.resolve(__dirname, enginePath);
    }

    const engineDirectory = path.dirname(enginePath);

    // On Windows, batch files (.bat, .cmd) must be spawned with shell: true
    const isBatchFile = process.platform === 'win32' && /\.(bat|cmd)$/i.test(enginePath);
    const spawnOptions = { cwd: engineDirectory };
    let command = enginePath;
    if (isBatchFile) {
      spawnOptions.shell = true;
      // Quote the path to handle spaces when shell: true is used
      command = `"${enginePath}"`;
    }

    engineProcess = spawn(command, [], spawnOptions);

    engineProcess.on('error', (err) => {
      console.error(`[${new Date().toISOString()}] Failed to start engine process. ${err.message}`);
      // Differentiate specific errors if needed for logs, but keep client message generic
      let clientMsg = "WRAPPER_ERROR: Failed to start engine process.";
      if (err.code === 'ENOENT') {
         clientMsg = "WRAPPER_ERROR: Engine executable not found.";
      }
      socket.write(clientMsg + '\n');
      cleanup();
    });

    // If the process fails to start, the 'error' event will be emitted and handled above.
    // We should not proceed if the process is not valid.
    if (engineProcess.pid === undefined) {
        return;
    }
    
    // Store options for the upper scope listener to use
    socket.engineOptions = engineDef.options;

    console.log(`[${new Date().toISOString()}] Started engine process: ${enginePath} (PID: ${engineProcess.pid})`);
    engineStarted = true;

    // Pipe all output from engine back to client
    engineProcess.stdout.on('data', (data) => {
      const output = data.toString().trim();
      
      // Reduce logging noise: Skip 'info' commands
      if (!output.startsWith("info")) {
        console.log(`[Engine -> Client] ${output}`);
      }
      if (socket.writable) {
        socket.write(data);
      }
    });
    engineProcess.stderr.on('data', (data) => {
      console.error(`[Engine ERROR] ${data.toString().trim()}`);
      if (socket.writable) {
        socket.write(data);
      }
    });

    engineProcess.on('close', (code) => {
      console.log(`[${new Date().toISOString()}] Engine process exited with code ${code}.`);
      engineStarted = false;
      // Ensure socket is closed when engine exits
      cleanup();
    });
  });

  socket.on('end', () => {
    console.log(`[${new Date().toISOString()}] Client sent FIN packet.`);
    cleanup();
  });

  socket.on('close', () => {
    console.log(`[${new Date().toISOString()}] Client disconnected.`);
    cleanup();
  });

  socket.on('error', (err) => {
    console.error(`[${new Date().toISOString()}] Socket error:`, err);
    cleanup();
  });
});

server.listen(PORT, HOST, () => {
  console.log(`[${new Date().toISOString()}] Single-port engine wrapper server listening on ${HOST}:${PORT}`);
  
  const enginesJsonPath = path.join(__dirname, 'engines.json');
  if (fs.existsSync(enginesJsonPath)) {
    console.log(`[${new Date().toISOString()}] engines.json found at ${enginesJsonPath}`);
    try {
      const content = fs.readFileSync(enginesJsonPath, 'utf-8');
      const engines = JSON.parse(content);
      console.log(`[${new Date().toISOString()}] Loaded ${engines.length} engines from engines.json:`);
      engines.forEach(e => console.log(`  - ${e.id}: ${e.name} (${e.path})`));
    } catch (e) {
      console.error(`[${new Date().toISOString()}] Failed to parse engines.json: ${e.message}`);
    }
  } else {
    console.error(`[${new Date().toISOString()}] engines.json not found. Please create one based on engines.json.example.`);
  }
});

// Graceful shutdown
const gracefulShutdown = (signal) => {
  console.log(`[${new Date().toISOString()}] Received ${signal}. Shutting down gracefully.`);
  server.close(() => {
    console.log(`[${new Date().toISOString()}] All connections closed. Server is shut down.`);
    process.exit(0);
  });

  // Forcefully close server after a timeout
  setTimeout(() => {
    console.error(`[${new Date().toISOString()}] Could not close connections in time, forcefully shutting down.`);
    process.exit(1);
  }, 10000); // 10 seconds
};

process.on('SIGINT', () => gracefulShutdown('SIGINT'));
process.on('SIGTERM', () => gracefulShutdown('SIGTERM'));
